src/audio/copier/copier.c

```diff
@@ -1502,8 +1502,6 @@ static int copier_params(struct comp_dev *dev, struct sof_ipc_stream_params *par
 					/* set up callback */
 					notifier_register(dev, cd->hd->chan,
 							  NOTIFIER_ID_DMA_COPY, copier_dma_cb, 0);
-
-				cd->hd->process = cd->converter[IPC4_COPIER_GATEWAY_PIN];
```

Comment:
preivous hd->process is based on input and endpoint buffer, after remove endpoint buffer, we have to choose converter[0], since it is based on audio input format and output format.

Reply:
This is the code in host-zephyr.c
```
/* set processing function */
if (params->direction == SOF_IPC_STREAM_CAPTURE)
	hd->process = pcm_get_conversion_function(host_buf_c->stream.frame_fmt,
						  dma_buf_c->stream.frame_fmt);
else
	hd->process = pcm_get_conversion_function(dma_buf_c->stream.frame_fmt,
						  host_buf_c->stream.frame_fmt);
```

host_buf_c is copier sink buffer for playback and copier source for capture. Just container sizes of source and sink formats are used here to select conversion function. And that is a correct behavior for host component.

For 24/32 to 16/16 conversion the code above correctly selects pcm_convert_s32_to_s16(), as 24/32 coming from driver has sample data in upper 24 bits of container, so should be treaded kind of 32/32.

Without fix, that correct pcm_convert_s32_to_s16() is replaced by pcm_convert_s24_to_s16() which is incorrect for host component. That would be correct for non-host component buffer to buffer conversion, but not for audio coming from/going to driver.

Reply:
@RanderWang can you please comment? The converter functions set up in host-zephyr.c is meant for IPC3 isnt it?

Reply:
@serhiy-katsyuba-intel I think this will break a bunch of QB tests. This particular piece of code was specifically added because the default converter function does not work as intended. Maybe the ipc4 converter function has a bug that needs to be fixed instead of removing this completely

Reply:
@ranj063 , seems indeed pcm_func_vc_map[] (used by copier pcm_get_conversion_vc_function()) is incorrect for ipc4_gtw_host and 16 to/from 24 bit conversion.
However, pcm_func_map (used by pcm_get_conversion_function() in host-zephyr) is correct and sufficient for all possible Host gateway formats.
Are those QB tests in CI? Maybe the failures you previously observed were caused by another bug, e.g. https://github.com/thesofproject/sof/pull/7477.

Reply:
@serhiy-katsyuba-intel no you can check the QB failures in this PR too

Reply:
> @RanderWang can you please comment? The converter function set up in host-zephyr.c is meant for IPC3 isnt it?

yes,  the converter function in host-zephyr.c is for IPC3 and it only supports c16/s16 (container size/ sample valid size), c32/s24,  c32/s32. For host gtw we need to support c24/s24 (supported by windows)